google/cloud/storage/internal/tuple_filter.h

```diff
@@ -0,0 +1,146 @@
+ * generate and a "filtering" member function, which either returns an empty
+ * tuple or a tuple containing the argument.
+ *
+ * @tparam T of the filtered element
+ * @tparam Filter whether to filter the element.
+ */
+template <typename T, bool Filter>
+struct TupleFilterItem {};
+
+/**
+ * Implementation of TupleFilterItem - true branch.
+ */
+template <typename T>
+struct TupleFilterItem<T, true> {
+  using Result = std::tuple<T>;
+  Result operator()(T&& t) const { return Result(std::forward<T>(t)); }
+};
+
+/**
+ * Implementation of TupleFilterItem - false branch.
+ */
+template <typename T>
+struct TupleFilterItem<T, false> {
+  using Result = std::tuple<>;
+  Result operator()(T&&) const { return Result(); }
+};
+
+/**
+ * A helper to compute the return type of `StaticTupleFilter`.
+ *
+ * @tparam TPred a type predicate telling if an element stays or is filtered out
+ * @tparam Tuple the type of the tuple to filter elements from
+ */
+template <template <class> class TPred, typename Tuple>
+struct FilteredTupleReturnType {};
+
+/**
+ * Implementation of FilteredTupleReturnType - recursive case.
+ *
+ * @tparam TPred a type predicate telling if an element stays or is filtered out
+ * @tparam Head the type of the first tuple's element
+ * @tparam Tail the remaining types tuple's elements
+ */
+template <template <class> class TPred, typename Head, typename... Tail>
+struct FilteredTupleReturnType<TPred, std::tuple<Head, Tail...>> {
+  using Result = typename std::conditional<
+      TPred<Head>::value,
+      typename google::cloud::internal::invoke_result<
+          decltype(std::tuple_cat<std::tuple<Head>,
+                                  typename FilteredTupleReturnType<
+                                      TPred, std::tuple<Tail...>>::Result>),
+          std::tuple<Head>,
+          typename FilteredTupleReturnType<TPred,
+                                           std::tuple<Tail...>>::Result>::type,
+      typename FilteredTupleReturnType<TPred,
+                                       std::tuple<Tail...>>::Result>::type;
+};
+
+/**
+ * Implementation of FilteredTupleReturnType - recursion end / empty tuple.
+ *
+ * @tparam TPred a type predicate telling if an element stays or is filtered out
+ */
+template <template <class> class TPred>
+struct FilteredTupleReturnType<TPred, std::tuple<>> {
+  using Result = std::tuple<>;
+};
+
+/**
+ * Filter elements from a tuple based on their type.
+ *
+ * A new tuple is returned with only the elements whose type satisfied the
+ * provided type predicate.
+ *
+ * @tparam TPred a type predicate telling if an element stays or is filtered out
+ * @tparam Args the type of the tuple's elements
+ *
+ * @param tuple the tuple to filter elements from
+ */
+template <template <class> class TPred, typename... Args>
+typename FilteredTupleReturnType<TPred, std::tuple<Args...>>::Result
+StaticTupleFilter(std::tuple<Args...> t) {
```

Comment:
This parameter, `t`, is being passed by value. is that intended?

Reply:
It wasn't, now it should be better.